File: Data Structures & Algorithms/meeting-schedule-ii/submission-5.cpp
/**
 * Definition of Interval:
 * class Interval {
 * public:
 *     int start, end;
 *     Interval(int start, int end) {
 *         this->start = start;
 *         this->end = end;
 *     }
 * }
 */

class Solution {
public:
    int minMeetingRooms(vector<Interval>& intervals) {
        //bucketing
        //min number of buckets
        /*
            (1,10),(2,3),(4,5),(6,7),(8,9)

            buckets = [40,10,]
        */
        int n = intervals.size();
        if(n == 0){
            return 0;
        }

        vector<int>buckets;

        sort(intervals.begin(),intervals.end(),[](const Interval& i1,const Interval& i2){
            if(i1.start == i2.start){
                return i1.end < i2.end;
            }
            return i1.start < i2.start;
        });

        buckets.push_back(intervals[0].end);
        for(int i=1; i<n; i++){
            int curr_arrival = intervals[i].start;
            int curr_leaving = intervals[i].end;
            bool added = false;

            for(int j=0; j<buckets.size(); j++){
                if(curr_arrival >= buckets[j]){
                    added = true;
                    buckets[j] = max(buckets[j],curr_leaving);
                    break;
                }
            }

            if(!added){
                buckets.push_back(curr_leaving);
            }
        }

        return buckets.size();
    }
};

File: Data Structures & Algorithms/meeting-schedule/submission-5.py
"""
Definition of Interval:
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end
"""

class Solution:
    def canAttendMeetings(self, intervals: List[Interval]) -> bool:
        """
        (0,30), (30,40), (40,60)
        """

        intervals = sorted(intervals, key= lambda a: (a.start,a.end))
        
        if len(intervals) == 0:
            return True

        last_interval = intervals[0]
        for i in range(1,len(intervals)):
            if last_interval.end > intervals[i].start:
                return False
            
            else:
                last_interval = intervals[i]

        return True